src/handle/command.ml

```diff
@@ -407,31 +407,79 @@ let handle : (Path.t -> Sign.t) -> sig_state -> p_command ->
             (* Just ignore the command with a warning. *)
             wrn pe.pos "Proof aborted."; ss
         | _ ->
-            (* Check that no metavariable remains. *)
-            if LibTerm.Meta.has true a then
-              (fatal_msg "The type of %a has unsolved metavariables.\n"
-                 pp_uid id;
-               fatal pe.pos "We have %a : %a." pp_uid id pp_term a);
-            (match t with
-             | Some(t) when LibTerm.Meta.has true t ->
-                 fatal_msg
-                   "The definition of %a has unsolved metavariables.\n"
-                   pp_uid id;
-                 fatal pe.pos "We have %a : %a ≔ %a."
-                   pp_uid id pp_term a pp_term t
-             | _ -> ());
             match pe.elt with
             | P_proof_abort -> assert false (* Handled above *)
             | P_proof_admit ->
                 (* If the proof is finished, display a warning. *)
                 if finished ps then
                   wrn pe.pos "The proof is finished. Use 'end' instead.";
+                (* Get a list of metas, ordered by dependency. *)
+                let metas =
+                  let type_m = LibTerm.Meta.get true a in
+                  let metas =
+                    match t with
+                    | Some t -> MetaSet.union type_m (LibTerm.Meta.get true t)
+                    | None -> type_m
+                  in
+                  let dep m_a m_b =
+                    (* If [m_a] depends on [m_b], then it is greater than
+                       [m_b], else they are equal. *)
+                    if LibTerm.Meta.depends m_a m_b then 1 else 0
```

Comment:
dep does not satisfy the conditions required for a comparison function. See http://caml.inria.fr/pub/docs/manual-ocaml/libref/Array.html.
```
 The following must be true for all x, y, z in a :
    cmp x y > 0 if and only if cmp y x < 0
    if cmp x y >= 0 and cmp y z >= 0 then cmp x z >= 0
```

Reply:
These properties should be guaranteed now, if I'm not mistaken.